crypto/evp/pmeth_gn.c

```diff
@@ -150,23 +150,53 @@ int EVP_PKEY_CTX_get_keygen_info(EVP_PKEY_CTX *ctx, int idx)
     return ctx->keygen_info[idx];
 }
 
+static EVP_PKEY *evp_pkey_new_key(int priv, int type, ENGINE *e,
+                                  const unsigned char *key, int keylen)
+{
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+
+    /* Note: for compatibility reasons we allow keylen to be < 0 in some cases */
+
+    ctx = EVP_PKEY_CTX_new_id(type, e);
+    if (ctx == NULL)
+        return NULL;
+    if (EVP_PKEY_keygen_init(ctx) <= 0)
+        goto err;
+    if (priv) {
+        if (EVP_PKEY_CTX_set_priv_key(ctx, key, (int)keylen) <= 0)
+            goto err;
+    } else {
+        if (EVP_PKEY_CTX_set_pub_key(ctx, key, (int)keylen) <= 0)
+            goto err;
+    }
+    if (EVP_PKEY_keygen(ctx, &pkey) <= 0)
+        goto err;
```

Comment:
It's currently not clear to me why this is called. If we pass it a key, why do we generate a key? Shouldn't the operation be either KEY_OP_PUBLIC or KEY_OP_PRIVATE?

Reply:
So it seems you turned the KEY_OP_KEYGEN parameter in something that doesn't generate a key anymore if a public or private key is already set and acts like the function was called with KEY_OP_PUBLIC or KEY_OP_PRIVATE instead. This seems to be rather unexpected behavior change of EVP_PKEY_keygen() to me and might break existing applications.

Reply:
A couple of important points here: Firstly I'm not breaking anything or changing any behaviour here. This is *existing* behaviour. This is exactly the way the existing "new mac key" stuff does things, and that is implemented across multiple algorithms (including in 1.1.0) so we can't really change this. All I have done here is created a synonym for creating "new private keys" because all the algorithms we have done this for so far are MACs - this is the first one where that isn't the case: but there's nothing special or specific about MACs - its just about setting raw private key data. I have extended the concept a little to bring in public keys too, but the underlying concept is the same.

Secondly, key generation typically consists of two steps: setting up the parameters for the key, and setting the private/public key data. Usually the key data is randomly generated but in the old (existing) "new mac key" way of doing things key generation effectively reduces down to setting up the parameters, and then setting the key data to what has been supplied. I'm just reusing that approach here. In a sense "keygen" is a bit of a misnomer - think of it more as "preparing a key for use".